lib/src/client/client_keepalive.dart

```diff
@@ -0,0 +1,241 @@
+  final Duration? pingInterval;
+
+  /// How long the connection should wait before shutting down after no response
+  /// to a ping.
+  ///
+  /// `GRPC_ARG_KEEPALIVE_TIMEOUT_MS` in the docs.
+  final Duration timeout;
+
+  /// If a connection with no active calls should be kept alive by sending
+  /// pings.
+  ///
+  /// `GRPC_ARG_KEEPALIVE_PERMIT_WITHOUT_CALLS` in the docs.
+  final bool permitWithoutCalls;
+
+  const ClientKeepAliveOptions({
+    this.pingInterval,
+    this.timeout = const Duration(milliseconds: 20000),
+    this.permitWithoutCalls = false,
+  });
+
+  bool get shouldSendPings => pingInterval != null;
+}
+
+sealed class KeepAliveState {
+  void disconnect();
+}
+
+/// Transport has no active rpcs. We don't need to do any keepalives.
+final class Idle extends KeepAliveState {
+  final Timer? pingTimer;
+  final Stopwatch timeSinceFrame;
+
+  Idle([this.pingTimer, Stopwatch? stopwatch])
+      : timeSinceFrame = stopwatch ?? clock.stopwatch()
+          ..start();
+
+  @override
+  void disconnect() {
+    pingTimer?.cancel();
+  }
+}
+
+/// We have scheduled a ping to be sent in the future. We may decide to delay
+/// it if we receive some data.
+final class PingScheduled extends KeepAliveState {
+  final Timer pingTimer;
+  final Stopwatch timeSinceFrame;
+
+  PingScheduled(this.pingTimer, [Stopwatch? stopwatch])
+      : timeSinceFrame = stopwatch ?? clock.stopwatch()
+          ..start();
+
+  @override
+  void disconnect() => pingTimer.cancel();
+}
+
+/// We need to delay the scheduled keepalive ping.
+final class PingDelayed extends KeepAliveState {
+  final Timer pingTimer;
+  final Stopwatch timeSinceFrame;
+
+  PingDelayed(this.pingTimer) : timeSinceFrame = clock.stopwatch()..start();
+
+  @override
+  void disconnect() => pingTimer.cancel();
+}
+
+/// The ping has been sent out. Waiting for a ping response.
+final class ShutdownScheduled extends KeepAliveState {
+  final bool isIdle;
+  final Timer shutdownTimer;
+
+  ShutdownScheduled(this.shutdownTimer, this.isIdle);
+
+  @override
+  void disconnect() => shutdownTimer.cancel();
+}
+
+final class Disconnected extends KeepAliveState {
+  @override
+  void disconnect() {}
+}
+
+enum _KeepAliveEvent {
+  onTransportActive,
+  onFrameReceived,
+  onTransportIdle,
+  sendPing,
+}
+
+/// A keep alive "manager", deciding when to send pings or shutdown based on the
+/// [ClientKeepAliveOptions].
+class ClientKeepAlive {
+  @visibleForTesting
+  KeepAliveState state = Idle();
+
+  final void Function() onPingTimeout;
+  final void Function() ping;
+
+  final ClientKeepAliveOptions _options;
+  Duration get _pingInterval => _options.pingInterval ?? Duration(days: 365);
+  final Map<Type, KeepAliveState? Function(KeepAliveState, _KeepAliveEvent)>
+      _stateTransform = {};
+
+  ClientKeepAlive({
+    required ClientKeepAliveOptions options,
+    required this.ping,
+    required this.onPingTimeout,
+  }) : _options = options {
+    _addEventHandler<Idle>((state, event) {
```

Comment:
Any specific reason that these is not just a method on the state itself. Using `runtimeType` + map for state machine feels a bit too dynamic.

Reply:
I didn't know how to wire through needed properties such as `ping` or `options` from `ClientKeepAlive` without making the code ugly or complicated. I thought this was a nice alternative.

Reply:
I am open for other ideas though :)